File: python/tests/test_module4_reduced_ae.py
from __future__ import annotations

import csv
import sys
import unittest
from pathlib import Path

import numpy as np

PYTHON_ROOT = Path(__file__).resolve().parents[1]
REPOSITORY_ROOT = PYTHON_ROOT.parent
SRC = PYTHON_ROOT / "src"
sys.path.insert(0, str(SRC))

from kinetic_mhd_pinn.module4_alfven_continuum_eigenmodes import (  # noqa: E402
    assemble_operator,
    build_profiles,
    convergence_study,
    load_case,
    solve_benchmark,
)


class ReducedAEBenchmarkTests(unittest.TestCase):
    @classmethod
    def setUpClass(cls) -> None:
        cls.case_path = PYTHON_ROOT / "cases" / "reduced_ae_benchmark" / "case.json"
        cls.case = load_case(cls.case_path)
        cls.result = solve_benchmark(cls.case)

    def test_case_files_are_identical_across_languages(self) -> None:
        matlab_case = REPOSITORY_ROOT / "matlab" / "cases" / "reduced_ae_benchmark" / "case.json"
        self.assertEqual(self.case_path.read_bytes(), matlab_case.read_bytes())

    def test_profiles_are_physical_and_crossing_is_resolved(self) -> None:
        p = self.result.profiles
        self.assertTrue(np.all(p.mass_density_kg_m3 > 0.0))
        self.assertTrue(np.all(p.alfven_speed_m_s > 0.0))
        self.assertGreater(p.crossing_s, 0.0)
        self.assertLess(p.crossing_s, 1.0)
        self.assertLess(p.gap_lower_rad_s, p.gap_upper_rad_s)
        self.assertTrue(np.all(p.local_lower_rad_s > 0.0))
        self.assertLess(p.gap_lower_rad_s / (2.0 * np.pi), self.result.frequency_hz[1])
        self.assertLess(self.result.frequency_hz[1], p.gap_upper_rad_s / (2.0 * np.pi))

    def test_operator_is_symmetric_and_positive(self) -> None:
        a = self.result.operator
        asymmetry = np.linalg.norm((a - a.T).toarray()) / np.linalg.norm(a.toarray())
        self.assertLess(asymmetry, self.case.verification.matrix_symmetry_tolerance)
        self.assertTrue(np.all(self.result.omega2_s2 > 0.0))

    def test_eigenvectors_are_normalized_and_residuals_are_small(self) -> None:
        h = self.result.profiles.h
        for index in range(self.result.eigenvectors.shape[1]):
            norm = h * np.dot(self.result.eigenvectors[:, index], self.result.eigenvectors[:, index])
            self.assertAlmostEqual(float(norm), 1.0, places=11)
        self.assertLess(float(np.max(self.result.residual_norms)), self.case.verification.eigen_residual_tolerance)

    def test_expected_alfven_scalings(self) -> None:
        doubled_b = solve_benchmark(
            self.case.with_updates(magnetic_field_t=2.0 * self.case.geometry.magnetic_field_t)
        )
        density_x4 = solve_benchmark(self.case.with_updates(density_scale=4.0))
        np.testing.assert_allclose(doubled_b.frequency_hz / self.result.frequency_hz, 2.0, rtol=2.0e-10, atol=0.0)
        np.testing.assert_allclose(density_x4.frequency_hz / self.result.frequency_hz, 0.5, rtol=2.0e-10, atol=0.0)

    def test_decoupled_limit_closes_avoided_crossing(self) -> None:
        decoupled = self.case.with_updates(coupling_fraction=0.0)
        profiles = build_profiles(decoupled)
        np.testing.assert_allclose(
            profiles.local_lower_rad_s,
            np.minimum(profiles.omega_m_rad_s, profiles.omega_mp1_rad_s),
            rtol=0.0,
            atol=1.0e-6,
        )
        np.testing.assert_allclose(
            profiles.local_upper_rad_s,
            np.maximum(profiles.omega_m_rad_s, profiles.omega_mp1_rad_s),
            rtol=0.0,
            atol=1.0e-6,
        )
        operator = assemble_operator(decoupled, profiles)
        n = profiles.s.size
        self.assertEqual(operator[:n, n:].nnz, 0)
        self.assertEqual(operator[n:, :n].nnz, 0)

    def test_grid_convergence_improves(self) -> None:
        rows = convergence_study(self.case)
        errors = [row["mode_1_relative_error_to_finest"] for row in rows[:-1]]
        self.assertGreater(errors[0], errors[1])
        self.assertGreater(errors[1], errors[2])
        self.assertLess(errors[2], 1.0e-4)

    def test_reference_regression(self) -> None:
        path = REPOSITORY_ROOT / "data" / "reference" / "reduced_ae_benchmark" / "eigenvalues.csv"
        with path.open(newline="", encoding="utf-8") as stream:
            rows = list(csv.DictReader(stream))
        reference = np.array([float(row["frequency_hz"]) for row in rows])
        np.testing.assert_allclose(
            self.result.frequency_hz,
            reference,
            rtol=self.case.verification.regression_eigenfrequency_relative_tolerance,
            atol=0.0,
        )

    def test_language_neutral_output_contract_is_complete(self) -> None:
        output = REPOSITORY_ROOT / "data" / "reference" / "reduced_ae_benchmark"
        required = {
            "profiles_and_continuum.csv",
            "eigenvalues.csv",
            "eigenmodes.csv",
            "convergence.csv",
            "metadata.json",
            "continuum.png",
            "continuum.pdf",
            "representative_eigenmode.png",
            "representative_eigenmode.pdf",
            "convergence.png",
            "convergence.pdf",
        }
        self.assertTrue(required.issubset({path.name for path in output.iterdir()}))


if __name__ == "__main__":
    unittest.main()

File: matlab/tests/test_module4_reduced_ae.m
function tests = test_module4_reduced_ae
tests = functiontests(localfunctions);
end

function setupOnce(testCase)
matlabRoot = fileparts(fileparts(mfilename('fullpath')));
repoRoot = fileparts(matlabRoot);
addpath(fullfile(matlabRoot, 'src'));
casePath = fullfile(matlabRoot, 'cases', 'reduced_ae_benchmark', 'case.json');
testCase.TestData.matlabRoot = matlabRoot;
testCase.TestData.repoRoot = repoRoot;
testCase.TestData.casePath = casePath;
testCase.TestData.caseData = kinetic_mhd_pinn.module4_alfven_continuum_eigenmodes.load_case(string(casePath));
testCase.TestData.result = kinetic_mhd_pinn.module4_alfven_continuum_eigenmodes.solve_benchmark(testCase.TestData.caseData);
end

function testCaseFilesMatch(testCase)
pythonCase = fullfile(testCase.TestData.repoRoot, 'python', 'cases', 'reduced_ae_benchmark', 'case.json');
verifyEqual(testCase, fileread(testCase.TestData.casePath), fileread(pythonCase));
end

function testOperatorSymmetryAndResidual(testCase)
result = testCase.TestData.result;
asymmetry = norm(result.operator - result.operator.', 'fro') / norm(result.operator, 'fro');
verifyLessThan(testCase, asymmetry, result.case_data.verification.matrix_symmetry_tolerance);
verifyGreaterThan(testCase, min(result.omega2_s2), 0.0);
verifyLessThan(testCase, max(result.residual_norms), result.case_data.verification.eigen_residual_tolerance);
end

function testNormalization(testCase)
result = testCase.TestData.result;
for j = 1:size(result.eigenvectors, 2)
    value = result.profiles.h * (result.eigenvectors(:,j).' * result.eigenvectors(:,j));
    verifyEqual(testCase, value, 1.0, 'AbsTol', 1.0e-10);
end
end

function testAlfvenScalings(testCase)
baseCase = testCase.TestData.caseData;
baseResult = testCase.TestData.result;
fieldCase = baseCase;
fieldCase.geometry.magnetic_field_t = 2.0 * baseCase.geometry.magnetic_field_t;
fieldResult = kinetic_mhd_pinn.module4_alfven_continuum_eigenmodes.solve_benchmark(fieldCase);
verifyEqual(testCase, fieldResult.frequency_hz ./ baseResult.frequency_hz, 2.0*ones(size(baseResult.frequency_hz)), 'RelTol', 1.0e-8);
densityCase = baseCase;
densityCase.profiles.mass_density_axis_kg_m3 = 4.0 * baseCase.profiles.mass_density_axis_kg_m3;
densityResult = kinetic_mhd_pinn.module4_alfven_continuum_eigenmodes.solve_benchmark(densityCase);
verifyEqual(testCase, densityResult.frequency_hz ./ baseResult.frequency_hz, 0.5*ones(size(baseResult.frequency_hz)), 'RelTol', 1.0e-8);
end

function testPythonReferenceEigenfrequencies(testCase)
referencePath = fullfile(testCase.TestData.repoRoot, 'data', 'reference', 'reduced_ae_benchmark', 'eigenvalues.csv');
reference = readtable(referencePath);
relative = abs(testCase.TestData.result.frequency_hz - reference.frequency_hz) ./ reference.frequency_hz;
verifyLessThan(testCase, max(relative), testCase.TestData.caseData.verification.python_matlab_eigenfrequency_relative_tolerance);
end

function testGridConvergence(testCase)
convergence = kinetic_mhd_pinn.module4_alfven_continuum_eigenmodes.convergence_study(testCase.TestData.caseData);
errors = convergence.mode_1_relative_error_to_finest(1:end-1);
verifyGreaterThan(testCase, errors(1), errors(2));
verifyGreaterThan(testCase, errors(2), errors(3));
verifyLessThan(testCase, errors(3), 1.0e-4);
end
